examples/matrix-factorisation/SparseRatingMatrix.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main where
+
+import Control.Monad (foldM, when)
+import MF
+import Torch.Autograd (grad, makeIndependent, toDependent)
+import Torch.DType (DType (Float))
+import Torch.Functional (matmul)
+import Torch.Internal.Managed.Type.Context (manual_seed_L)
+import Torch.NN
+  ( Randomizable,
+    sample,
+    sgd,
+  )
+import Torch.Tensor (Tensor, asTensor, toDouble, toInt)
+import Torch.TensorFactories (randint', randn')
+
+type Item = Int
+
+type User = Int
+
+type RatingValue = Double
+
+data Rating = Rating Item User RatingValue deriving (Show)
+
+data Ratings = Ratings [Rating]
+
+data RatingsSpec
+  = RatingsSpec {usersNumber :: Int, itemsNumber :: Int, ratingNumber :: Int}
+
+instance Randomizable RatingsSpec Ratings where
+  sample (RatingsSpec n m l) = do
+    ratings <- randn' [1, l]
+    users <- randint' 1 n [1, l]
+    items <- randint' 1 m [1, l]
+    let list_ratings = toDouble <$> (getEntryByIndex ratings 0) <$> [0 .. (l -1)]
+        list_users = toInt <$> (getEntryByIndex users 0) <$> [0 .. (l -1)]
+        list_items = toInt <$> (getEntryByIndex items 0) <$> [0 .. (l -1)]
+    pure $ Ratings [Rating u i r | (u, i, r) <- zip3 list_users list_items list_ratings]
+
+type TensorEntries = Double
+
+lossMF :: MatrixFact -> [Rating] -> Tensor
+lossMF (MatrixFact u v) rating_list =
+  let uv = mulMF (MatrixFact u v)
+      mse_loss' sum (Rating user item rating) = sum + (uv_ui - tensor_rating) ^ 2
+        where
+          uv_ui = getEntryByIndex uv user item
+          tensor_rating = asTensor rating
+   in foldl mse_loss' (asTensor (0 :: TensorEntries)) rating_list
```

Comment:
once you have a way to compute `uv` only for the `users` and `items` in `RatingBatch`, you can use the standard `mse_loss` function to calculate the l2 norm distance.

Reply:
What do you think about this?
```haskell
lossMF :: MatrixFact -> [Rating] -> Tensor
lossMF (MatrixFact u v) rating_list =
    let
      mse_loss' sum (Rating user item rating) = sum + (uv_ui - tensor_rating) ^ 2
        where
          uv_ui = matmul (u' @@ (user,()) ) (v' @@ ((),item) )
          u' = toDependent u
          v' = toDependent v
          tensor_rating = asTensor rating
   in foldl mse_loss' (asTensor (0 :: TensorEntries)) rating_list
```

Reply:
I don't still see a clear of way of doing it. I would do it in this way:
1. making a `Double` list including the desired entries of `uv`. 
2. using `asTensor` to make it a tensor
3. looking at the `mse_loss` of the latter Tensor and the ratings.
I suspect if we can take gradients from it (since we use `toDouble` and `asTensor`).